+ update ProgrammeUpdater::ExtendedWebClient to properly access legac…

neoh4x0r opened the pull request:
I just pushed a commit to grab the releases from github rather than nexusmods
and it should fix issue: #346

It seems that legacy-api urls require a user-agent of : "Nexus Client v0.65.5" (or later).
This works for any legacy-api that does not require a login.

However, to use an ordinary webclient (with proper user-agent) you must pass the login authentication token as a cookie; although it would better to use the repository api with channel/proxy factory.

But, releases are now obtained from github using their release api:
https://api.github.com/repos/Nexus-Mods/Nexus-Mod-Manager/releases

neoh4x0r committed:
+ update ProgrammeUpdater::ExtendedWebClient to properly access legacy-api urls without recieving a 403 (Forbidden) status code
+ update ProgrammeUpdater::GetNewProgrammeVersion(out string p_strDownloadUri)
  ** fix processing of the version and download url

neoh4x0r committed:
+ update ProgrammeUpdater to grab releases from GitHub using their releases api and a json parser
https://api.github.com/repos/Nexus-Mods/Nexus-Mod-Manager/releases

neoh4x0r committed:
+ make a few changes in the previous commit that would have caused some compliation errors

neoh4x0r committed:
+ rename GitHubReleases back to Releases

neoh4x0r committed:
+ in GithubReleaseParser::ParseReleaseInformation
  ** the method should return true only if we have 1 or more releases
     instead of returning true in any case

neoh4x0r committed:
+ update ParseReleaseInformation to avoid null reference exceptions

squid-box reviewed (commented):
Overall I think this is a better solution than my PR (I will close it). The only real comment I have is that I'm not a big fan of having several classes within the same file - but it's not a big objection since that's already the case in many files in NMM already.

squid-box committed:
Removed a rogue ')' causing a compilation error.

neoh4x0r commented:
Yeah I am also not a big fan of having multiple classes in the same file.
However, I chose to do that because those classes are only used in the one location (as there shouldn't be any reason to expose them outside of it).

Also to ensure that we have the latest release we should probably change the url:
from:
https://api.github.com/repos/Nexus-Mods/Nexus-Mod-Manager/releases
to:
https://api.github.com/repos/Nexus-Mods/Nexus-Mod-Manager/releases/latest

Then the **ParseReleaseInformation** would not be arbitrarily assuming that the latest release is the first element in the list.

EDIT: I just pushed changes so that we do not have to assume the above.

neoh4x0r committed:
+ change the latest release url to:
     https://api.github.com/repos/Nexus-Mods/Nexus-Mod-Manager/releases/latest

  The previous code would use parse all releases and it would assume that
  the latest one was the first in the list....now we don't have to assume that.
